File: algorithms/cracking_the_coding_interview/hash_tables_ice_cream_parlor.php
<?php
declare(strict_types=1);

/**
 * @param int[] $flavors
 * @param int $money
 */
function solve(array $flavors, int $money)
{
    $map = [];
    $result = [];

    foreach ($flavors as $index => $flavor) {
        if (!empty($map[$flavor])) {
            $result = [$index + 1, $map[$flavor]];
            break;
        }

        $map[$money - $flavor] = $index + 1;
    }

    sort($result);
    echo $result[0] . ' ' . $result[1] . "\n";
}

$handle = fopen('php://stdin', 'r');
fscanf($handle, '%i', $t);
for ($i = 0; $i < $t; $i++) {
    fscanf($handle, '%i', $money);
    fscanf($handle, '%i', $n);
    $flavors = array_map('intval', explode(' ', fgets($handle)));
    solve($flavors, $money);
}

File: algorithms/cracking_the_coding_interview/hash_tables_ice_cream_parlor.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

func solve(flavors []int, money int) {
	flavorMap := make(map[int]int)
	var result []int

	for index, flavor := range flavors {
		if value, ok := flavorMap[flavor]; ok {
			result = append(result, index+1, value)
			break
		}

		flavorMap[money-flavor] = index + 1
	}

	sort.Ints(result)
	fmt.Printf("%d %d \n", result[0], result[1])
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	visitsLine, _ := reader.ReadString('\n')
	visits, _ := strconv.Atoi(strings.TrimSpace(visitsLine))

	for i := 0; i < visits; i++ {
		moneyLine, _ := reader.ReadString('\n')
		money, _ := strconv.Atoi(strings.TrimSpace(moneyLine))

		reader.ReadString('\n')

		flavorsLine, _ := reader.ReadString('\n')
		flavorsLine = strings.TrimSpace(flavorsLine)
		flavorsString := strings.Fields(flavorsLine)
		flavors := make([]int, len(flavorsString))
		for i := range flavorsString {
			flavors[i], _ = strconv.Atoi(flavorsString[i])
		}

		solve(flavors, money)
	}
}
